File: Code/12_GPIO_Polling/Ex2.c
#include <stdio.h>
#include <fcntl.h>
#include <sys/poll.h>
#include <unistd.h>

int main(void)
{
	struct pollfd pfd;
	char buffer;
	system("echo 4       > /sys/class/gpio/export");
	system("echo falling > /sys/class/gpio/gpio4/edge");
	system("echo in      > /sys/class/gpio/gpio4/direction");
	pfd.fd = open("/sys/class/gpio/gpio4/value", O_RDONLY);
	if(pfd.fd < 0)
	{
		puts("Erro abrindo /sys/class/gpio/gpio4/value");
		puts("Execute este programa como root");
		return 1;
	}
	read(pfd.fd, &buffer, 1);
	puts("Pressione o botao...");
	pfd.events = POLLPRI | POLLERR;
	pfd.revents = 0;
	poll(&pfd, 1, -1);
	if(pfd.revents & POLLPRI) puts("POLLPRI detectado");
	if(pfd.revents & POLLERR) puts("POLLERR detectado");
	close(pfd.fd);
	system("echo 4 > /sys/class/gpio/unexport");
	return 0;
}

File: 3_Programação_RPi/3.3_GPIO_Polling/Python/Ex2.py
import RPi.GPIO as gpio
import os, select

pin = 4
os.system("echo %d      > /sys/class/gpio/export" % pin)
os.system("echo falling > /sys/class/gpio/gpio%d/edge" % pin)
os.system("echo in      > /sys/class/gpio/gpio%d/direction" % pin)

fd = os.open("/sys/class/gpio/gpio%d/value" % pin, os.O_RDONLY)
os.read(fd, 1)
pfd = select.poll()
pfd.register(fd, select.POLLPRI | select.POLLERR)

print("Pressione o botao...")
ret = pfd.poll()
if ret[0][1] & select.POLLPRI:
	print("POLLPRI detectado")
if ret[0][1] & select.POLLERR:
	print("POLLERR detectado")
print("Borda de descida!")
os.close(fd)
os.system("echo %d > /sys/class/gpio/unexport" % pin)
